### ISSUE TITLE
[BUG] Docker-proxy binds on ports but no container is running (anymore)

### ISSUE THREAD

**Issue by nagua** [moby/moby#25981]
**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:02:53 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:02:53 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 6
 Running: 0
 Paused: 0
 Stopped: 6
Images: 93
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 211
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.748 GiB
Name: motocom.2hg.org
ID: O4YH:4KEL:GEY3:C26M:QWMN:PCLK:RF6E:DQSG:XPWG:SSBE:FVE3:GWBL
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No kernel memory limit support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
This is a dedicated server running Debian Jessie. Everything up to date.

```
# docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

```
# netstat -ap
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 *:3389                  *:*                     LISTEN      890/openvpn
tcp        0      0 localhost:mysql         *:*                     LISTEN      1276/mysqld
tcp        0      0 *:ssh                   *:*                     LISTEN      727/sshd
tcp        0    336 xxx:ssh     ip1f10d8b9.dynami:60637 ESTABLISHED 1641/sshd: xxx
tcp        0      1 xxx:ssh     116.31.116.52:53774     FIN_WAIT1   -
tcp6       0      0 [::]:smtp               [::]:*                  LISTEN      6346/docker-proxy
tcp6       0      0 [::]:5280               [::]:*                  LISTEN      6366/docker-proxy
tcp6       0      0 [::]:imaps              [::]:*                  LISTEN      6316/docker-proxy
tcp6       0      0 [::]:5443               [::]:*                  LISTEN      6356/docker-proxy
tcp6       0      0 [::]:xmpp-client        [::]:*                  LISTEN      6386/docker-proxy
tcp6       0      0 [::]:submission         [::]:*                  LISTEN      6326/docker-proxy
tcp6       0      0 [::]:imap2              [::]:*                  LISTEN      6336/docker-proxy
tcp6       0      0 [::]:xmpp-server        [::]:*                  LISTEN      6376/docker-proxy
[...]
```

**Steps to reproduce the issue:**
1. I upgraded from version 1.12.0 to 1.12.1.
2. I tried to restart my containers but they failed to allocate ports.
3. The ports are allocated by `docker-proxy` but no container is running.

**Describe the results you received:**
I can not start the containers because the ports are already in use.

**Describe the results you expected:**
Docker-proxy does not allocate ports docker does not use.

**Additional information you deem important (e.g. issue happens only occasionally):**

I already tried to restrict docker to ipv4 but this was not changing anything. I also tried to restart docker/server but docker also allocates many ports. I use docker-compose to manage my containers if this matters.

Are there more information needed? Is there a quick workaround available make docker think that the ports are not needed anymore?

**Comment by amitkumarj441** [comment]
Hey @nagua,

It isn't really possible to simply close a port from outside the application that opened the socket listening on it. The only way to do this is to completely kill the process that owns the port. Then, in about a minute or two, the port will become available again for use. Here's what's going on (if you don't care, skip to the end where I show you how to kill the process owning a particular port):

Anyway, here's how to kill a process that owns a particular port:

`sudo netstat -ap | grep :<port_number>`

That will output the line corresponding to the process holding port . Then, look in the last column, you'll see /. Then execute this:

`kill  <pid>`

For more info: #5666 and #6682

**Comment by nagua** [comment]
I now moved /var/lib/docker away, restarted docker and copied the volumes directory back in the newly created docker folder. Now everything is working again. If you need some files from there for debug purposes, i have the old docker folder lying around.

**Comment by sanimej** [comment]
@nagua I am not able to recreate this. I had some containers running with port published. When I restart the daemon `docker-proxy` processes also go down and the ports are available. I didn't try the upgrade from 1.12.0 to 1.12.1. But it shouldn't be very different from a daemon restart. Did you notice when the `docker-proxy` processes moved to zombie state ?

**Comment by kossmoss** [comment]
Have same problem running docker (v 1.12.1) via docker-compose inside virtual machine running on Vagrant.
Noticed this problem several times after restarting virtual machine (probably this is sign of incorrect terminating docker daemon processes).
I have several occupied ports (and docker ps shows no any actual docker containers running at the same time)
Simple killing processes and restarting docker service (or even whole virtual machine) doesnt solve the problem. Every time docker daemon restarts, it brings those zombie docker-proxy processes again and ports are still unavailable for new containers.

I'm going to try to solve problem by saving virtual machine snapshot and after that will try to clear /var/lib/docker as @nagua did.

**Comment by cpuguy83** [comment]
@kossmoss Can you explain the "zombie docker-proxy processes" comment?
Do you see `docker-proxy` running while the daemon is not up?

**Comment by nagua** [comment]
@kossmoss When I started the docker daemon it starts up the docker-proxy processes. For ports where no container exists anymore that uses these ports. Trying to kill the docker-proxy processes doesn't help because they do not exit and ends up in a zombie process state. So they do not release the ports.

So I had no way to use the ports again while docker was running.

**Comment by hopeseekr** [comment]
We have this same problem. It's binding to ports 3306 and 5900 that containers usually bind, but now none of those containers will start. Help!

**Comment by jwodder** [comment]
We appears to be experiencing this same problem.

Docker 1.12.0
Ubuntu 16.04
Kernel: 4.4.0-36-generic

After restarting the Docker daemon (in order to fix a presumably unrelated problem with an unstoppable container) on an instance with 70 containers, approximately 44 of which have exposed ports, four of the containers were unable to start (even after deleting & recreating them) because their ports were in use by `docker-proxy` instances.  I tried killing one of the instances, but it just became a zombie, which Docker still hasn't reaped after about 10 minutes.

**Comment by kossmoss** [comment]
@cpuguy83 sorry for the late reply. Probably, "zombie" is not proper definition for that - I was limited in time and didn't performed too much of investigations and can't say how exactly we need to call those processes. My way to look on currently opened ports is **sudo netcat -tunlp** command or **sudo service docker status**.  In both cases, after docker daemon started, I see those docker-proxy processes occupying the ports, but in the same time see no any running docker containers in **docker ps** output.

When I kill docker-proxy process, it actually destroyed and port is not occupied until next docker daemon restart.

When docker daemon is stopped, there's no any running docker-proxy processes. They only running when docker daemon is running. Seems they are starting automatically at some point of docker daemon starting procedure. This also seems like docker holds those processes' state somewhere when stopped and brings them live when started, but does not link those processes to any running docker containers. 

Maybe this problem caused by starting containers via docker-compose instead of simple docker run? @hopeseekr are you using docker-compose?

**Comment by rdavaillaud** [comment]
Well, we've got the same problem, with compose 1.8, docker 1.12.1 on fedora 24.

**Comment by rdavaillaud** [comment]
Well, I managed to find a workaround:

- stop docker
- remove all internal docker network: rm /var/lib/docker/network/files/
- start docker

**Comment by sanimej** [comment]
@rdavaillaud @kossmoss I couldn't recreate the issue. But I wasn't using compose. All the reported occurrences seem to be happening with compose. Can you give the exact steps you are using that results in the issue ?

**Comment by chione** [comment]
I work with rdavaillaud and we have 4 pc using this docker and compose configuration, and this happend to only one of the pc one day after a computer reboot.

So we don't really know how to reproduce.

**Comment by rdavaillaud** [comment]
This is our docker-compose.yml, nothing fancy here I think, and no network definition.
```yaml
version: '2'
services:
  frontal:
    build:
      context: ./Docker/frontalWeb
    container_name: frontal
    extra_hosts:
     - "app.dev.local:127.0.0.1"
     - "www.app.dev.local:127.0.0.1"
     - "ws.app.dev.local:127.0.0.1"
    image: local/frontalweb
    links:
      - bdd:HBS-BDD
      - tilecache:carto.app.dev.local
      - tilecache:carto1.app.dev.local
      - tilecache:carto2.app.dev.local
      - tilecache:carto3.app.dev.local
      - tilecache:carto4.app.dev.local
    ports:
      - "80:80"
      - "443:443"
    privileged: true
    volumes:
      - /sys/fs/cgroup:/sys/fs/cgroup:ro
      - ./:/var/www/app/app-source
      - ./../app-datas:/var/local/app-datas
      - ./System/apache/:/etc/httpd/sites-enabled/
      - ./Docker/frontalWeb/docker-entrypoint.d:/docker-entrypoint.d
      - ./app.crt:/etc/ssl/app.com/app.com.cer
      - ./app.key:/etc/ssl/app.com/app.com.key
      - ./app.crt:/etc/ssl/app.com/app.crt

  tilecache:
    build:
      context: ./Docker/tilecache
    container_name: tilecache
    links:
      - mapserver:carto.app.dev.local
      - mapserver:carto1.app.dev.local
      - mapserver:carto2.app.dev.local
      - mapserver:carto3.app.dev.local
      - mapserver:carto4.app.dev.local
    ports:
      - "82:80"
    volumes:
      - ./Carto:/var/www/app/app/Carto
      - ./../mapfiles/app:/var/local/spatial/app
      - ./../app-datas:/var/local/app-datas
      - ./Docker/tilecache/docker-entrypoint.d:/docker-entrypoint.d

  mapserver:
    build:
      context: ./Docker/mapserver
    container_name: mapserver
    links:
      - bdd:HBS-BDD
    ports:
      - "81:80"
    volumes:
      - ./Carto:/var/www/app/app/Carto
      - ./../mapfiles/app:/var/local/spatial/app
      - ./../app-datas:/var/local/app-datas
      - ./Docker/mapserver/docker-entrypoint.d:/docker-entrypoint.d

  bdd:
    build:
      context: ./Docker/postgresql
    cap_add:
      - SYS_ADMIN
    container_name: bdd
    ports:
      - "5432:5432"
    privileged: true
    security_opt:
      - seccomp:unconfined
    volumes:
      - /sys/fs/cgroup:/sys/fs/cgroup:ro
      - ./DataBase:/var/www/app/app/DataBase
      - ./../app-datas-bdd:/var/local/app-datas
      - ./Docker/postgresql/docker-entrypoint.d:/docker-entrypoint.d
```

The docker containers are built from official centos7-systemd image.
The hosts are on Fedora 24, Docker-engine 1.12.1 and Compose 1.8.

We will tell you if it happens again.

**Comment by kossmoss** [comment]
@sanimej As I mentioned before I noticed the problem after restarting virtual machine. So I know only one obvious step to try to reproduce the problem - restart machine (or even just turn the power off) having running containers (mind the docker-compose - maybe it's also important). Probably, restart several times. It's kinda cruel way, so I can't recommend to do thatm but don't know any other way yet )

**Comment by glenpike** [comment]
removing /var/lib/docker/network/files/ worked for me.

**Comment by pcornelissen** [comment]
I'm having the same problem on docker for Mac.

```
docker info
Containers: 27
 Running: 4
 Paused: 0
 Stopped: 23
Images: 250
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 310
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.19-moby
Operating System: Alpine Linux v3.4
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 2.934 GiB
Name: moby
ID: 2STL:DVY7:P6IW:M2IO:I35K:MP3R:LCE7:NPJQ:IQYN:5OSY:GHLJ:6VUR
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 66
 Goroutines: 82
 System Time: 2016-09-08T16:36:04.267966976Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
```

But I have no /var/lib/docker to clean up :-(

**Comment by cpuguy83** [comment]
@sanimej Is this some change that we are storing port state information in bolt?

**Comment by dursk** [comment]
Just ran into this same issue w/ Docker for Mac, `v1.12.1` and docker-compose `v1.8`

**Comment by aboch** [comment]
This issue is a side-effect of the changes added to support the `--live-restore` feature.

From docker 1.12.0 onward, when the bridge network driver comes up, it will restores the bridge network endpoints it finds in the store.
While doing this, it also restores the port bindings associated with the endpoint, if any.

Note:
- Under normal condition at daemon boot, no endpoints are present in the store.
- If stale endpoints are present (this is usually the case of an ungraceful shutdown of the daemon with running containers), they are expected to be removed during boot as part of the stale sandbox cleanup process run by libnetwork core.
- If endpoints are present because of the live-restore, they will not be removed because the sandbox cleanup will not happen for the containers which are running.

The issue seems like the sandbox for stale endpoints from older docker version run is not present, therefore libnetwork core does not invoke the cleanup of the stale endpoints with the driver.

I believe the stale endpoints issue can be fixed by removing the networks and restarting the daemon. Because during the bridge endpoint restore, the endpoint is discarded and removed from store if the corresponding network is missing.

If the above does not work, one solution is to manually remove the problematic `docker/network/v1.0/bridge-endpoint/<id>` key value from the store. I just found out this cli tool to browse and modify a boltdb store, but did not have much luck using it so far (https://github.com/br0xen/boltbrowser).

Otherwise, last resort is to remove the `/var/lib/docker/network/files/local-kv.db` file, before starting the daemon.

On a side note, there is also a bug which will cause this issue. It is explained and fixed in https://github.com/docker/libnetwork/pull/1504 and will be available in next release.

**Comment by pcornelissen** [comment]
Hi!

I am not 100% sure because it happened a while ago, but I think I removed everything including the networks and the problem remained. But maybe I didn't restart afterwards or something like that. Nevertheless thanks for the analysis!

**Comment by justincormack** [comment]
cc @djs55

**Comment by Luzifer** [comment]
Confirmed temp-fix:

```bash
# systemctl stop docker; rm /var/lib/docker/network/files/local-kv.db; systemctl start docker
# docker ps -a | awk '{print $1}' | docker rm -f
```

Afterwards a clean env is available and containers are able to bind to their respective ports.

**Comment by aarnaud** [comment]
```
docker ps -a | awk '{print $1}' | docker rm -f
```
is it really necessary ? 

Like @rdavaillaud this workaround work for me:
```
systemctl stop docker
rm -rf /var/lib/docker/network/files
systemctl start docker
```

**Comment by jeanpralo** [comment]
Workaround of removing /var/lib/docker/network/files/local-kv.db works.

The problem it creates though is that if you happen to have containers sharing the same network then you will have to make sure you restart all of them :(

This is quite a big problem when you have a server crashing and you containers do not come back up  because of that. Any chance of a fix regarding what @aboch  found ? @thaJeztah

**Comment by ukaszg** [comment]
Coworker had this exact problem, changing fs driver aufs->devicemapper helped.

**Comment by marcelmfs** [comment]
I just saw this behaviour in 1.12.1, docker-compose 1.8.1.

We have some short lived containers that do bind to some ports for registration purposes, and after some runs, docker-proxy were still running referencing old runs of already stopped & removed containers, and then at some point docker reuses one of the IPs but the iptables rules are already in place and it causes all sorts of routing problems (trying to request one service reaches another).

This is really bad. It will happen in production (given that we expect docker to be a long lived process).

**Comment by cpuguy83** [comment]
@marcelmfs please update to 1.12.5.
The above mentioned bug should be fixed there (as of 1.12.3)

**Comment by marcelmfs** [comment]
@cpuguy thanks, I'll do it asap.

**Comment by rohitsakala** [comment]
Hi, I faced the same problem. My docker version is 1.12.0. Is this fixed. If it is, I will update my docker.

Thanks.

**Comment by amitkumarj441** [comment]
@rohitsakala Your Docker v1.12.0 is old and this problem is fixed in newest docker v1.13.x and it seems like you have docker-compose v1.8, if you want to continue with this Docker version then removing /var/lib/docker/network/files/ will works for your Docker version.

**Comment by rohitsakala** [comment]
@amitkumarj441 Thanks. I will update the docker as I need a permanent solution to it.

**Comment by thaJeztah** [comment]
Yes, let me close this issue because the issue reported here was resolved in 1.12.3

**Comment by Davor111** [comment]
I just had this exact issue happen to me on Docker version 1.13.1, build 092cba3, docker-compose version 1.10.0, build 4bd6f1a running Ubuntu 14.04 with just 5 containers running a low traffic Wordpress instance. For routing I was using an nginx image straight from docker hub with a couple of conf files for reverse proxying. I was able to restore the setup by using the method mentioned above (stopping docker all together, removing the network files and starting it all up again).

Honestly I'm a little surprised as this setup was running flawlessly for the past 6 months. The only recent change was the addition of certificates to the nginx reverse proxy. Please let me know, if I should share docker info, etc.

**Comment by Lykathia** [comment]
Just had this happen with Docker version 18.02.0-ce, build fc4de447b5

stop, removing the local-kv.db file, and starting again resolved the issue

**Comment by jstoja** [comment]
@Lykathia No need to delete the local-kv.db file.
1. kill the docker-proxy process owning the port
2. restart the docker daemon

**Comment by Lykathia** [comment]
@jstoja just had the same error today (this time on version 18.05.0-ce, build f150324782). Killing the processing and restarting did not address the issue.

Removing the local-kv.db file again did solve it.

The issue this time happened after a power outage in the server room. I'm guessing something didn't get cleaned up correctly.
/shrug

**Comment by Ghoughpteighbteau** [comment]
I just had this issue occur with me as well while testing out some swarm networking issues
```
Client:
 Version:      18.05.0-ce
 API version:  1.37
 Go version:   go1.10.2
 Git commit:   f150324782
 Built:        Wed May 16 22:27:45 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.05.0-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.10.2
  Git commit:   f150324782
  Built:        Wed May 16 22:28:17 2018
  OS/Arch:      linux/amd64
  Experimental: false
```
killing the process did not work. systemctl stop docker stopped the docker-proxy process and no longer bound to the port, when I started it back up it rebound to the port. I had to stop docker and remove local-kv.db file to prevent docker from pointlessly binding to that port.

**Comment by ihulsbus** [comment]
Issue occurs here too.

Output of `docker version`:
```
Client:
 Version:           18.06.1-ce
 API version:       1.30 (downgraded from 1.38)
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:24:51 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          17.06.2-ce
  API version:      1.30 (minimum version 1.12)
  Go version:       go1.8.3
  Git commit:       a04f55b
  Built:            Thu Sep 21 20:36:57 2017
  OS/Arch:          linux/amd64
  Experimental:     false
```

Output of `docker info`:
```
Containers: 4
 Running: 3
 Paused: 0
 Stopped: 1
Images: 4
Server Version: 17.06.2-ce
Storage Driver: aufs
 Root Dir: /var/snap/docker/common/var-lib-docker/aufs
 Backing Filesystem: extfs
 Dirs: 28
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6e23458c129b551d5c9871e5174f6b1b7f6d1170
runc version: 810190ceaa507aa2727d7ae6f4790c76ec150bd2
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-33-generic
Operating System: Ubuntu Core 16
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.466GiB
Name: server
ID: XM3Y:EDCJ:JXRQ:QET5:LEBD:JURD:NKZA:XABF:YHQR:DPC6:O7FD:6X5J
Docker Root Dir: /var/snap/docker/common/var-lib-docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 34
 Goroutines: 38
 System Time: 2018-09-08T18:43:13.438017572Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Comment by Ezwen** [comment]
I encountered something similar this morning: a container had exited and could not restart because its port was still in use by the `docker-proxy` process. I has to restart docker entirely to fix the issue.

```
# docker info
Containers: 27
 Running: 25
 Paused: 0
 Stopped: 2
Images: 29
Server Version: 18.06.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: journald
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.0-8-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.859GiB
Name: hallownest
ID: IFMK:N5V6:TADM:HFRT:4PEC:HMR5:2VJP:OOML:IA5W:USWR:EILG:SDS4
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Comment by Ezwen** [comment]
(note: I still encounter the problem regularly, and still have each time to restart the whole docker daemon to solve it)

**Comment by mellena1** [comment]
I have this same problem on a ubuntu 16.04 ec2 instance using docker-compose and `restart: always`. The container is fails on startup because it is unable to get IAM permissions (I'm assuming the container starts before the IAM stuff gets set up and starts working), and then tries to restart but fails because a docker-proxy process still has its port held.

Only fix is running `service docker restart`, but this is not ideal as in an AWS autoscaling group the service never becomes healthy.

**Comment by dragon9783** [comment]
same issue on docker 18.03.1-ce, remove all of container, restart docker daemon, then solved it

**Comment by EronWright** [comment]
I had this issue with port 8080, struggled to find what was listening on it.   I traced it to a process in the VM:
```
/ # netstat -plnt | grep 8080
tcp        0      0 :::8080                 :::*                    LISTEN      34227/api-server

/ # ps | grep 34227
34227 root      0:06 /api-server --kubeconfig  --authentication-kubeconfig= --authorization-kubeconfig= --service-name=compose-api --service-namespace docker --etcd-servers=https://127.0.0.1:2379 --etcd-cafile=/etc/docker-compose/etcd/ca.crt --etcd-certfile=/etc/docker-compose/etcd/client.crt --etcd-keyfile=/etc/docker-compose/etcd/client.key --secure-port=9443
```

I take this to mean that this error may indicate a conflict with the Kubernetes infrastructure that Docker Desktop provides.

**Comment by pcornelissen** [comment]
No, this happens without kubernetes as well

**Comment by djs55** [comment]
@EronWright the specific case of port 8080 in the recent release of Docker Desktop is caused by https://github.com/docker/compose-on-kubernetes/issues/70 and also reported here: https://github.com/docker/for-mac/issues/3522 . There's a workaround in the latest development builds of Docker Desktop if you'd like to try it: [Mac](https://download-stage.docker.com/mac/edge/31534/Docker.dmg) and [Win](https://download-stage.docker.com/win/edge/31536/Docker%20Desktop%20Installer.exe)

**Comment by Ezwen** [comment]
Small update: in fact I still encounter this problem even with `"userland-proxy": false`.

**Comment by TomerFi** [comment]
> Well, I've found a workaround:
> 
> * stop docker
> * remove all internal docker network: rm /var/lib/docker/network/files/
> * start docker
By @rdavaillaud did the trick on my end, thanks!

**Comment by bjeanes** [comment]
I am getting this basically weekly.

```
Client:
 Version:           18.09.7
 API version:       1.39
 Go version:        go1.10.1
 Git commit:        2d0083d
 Built:             Fri Aug 16 14:20:06 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.7
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.1
  Git commit:       2d0083d
  Built:            Wed Aug 14 19:41:23 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

Why is this marked closed?

**Comment by quentin-geru** [comment]
This happens to me on my laptop whenever I reboot it and try to launch docker-airflow which uses the 8080:
`Client: Docker Engine - Community
 Version:           19.03.5
 API version:       1.39 (downgraded from 1.40)
 Go version:        go1.12.12
 Git commit:        633a0ea838
 Built:             Wed Nov 13 07:29:52 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.9
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.13.4
  Git commit:       9552f2b
  Built:            Sat Nov 16 01:07:48 2019
  OS/Arch:          linux/amd64
  Experimental:     false
`

**Comment by Trinition** [comment]
I've started experiencing this as well, except it's not `docker-procy` holding the ports but `com.docker.backend`.

**Comment by thaJeztah** [comment]
@Trinition if that's on Docker Desktop (docker for mac / docker for windows), best to report that in the corresponding issue tracker: https://github.com/docker/for-mac/issues

**Comment by stuckj** [comment]
This just happened to me as well. Also using docker-compose (via portainer). So, libcompose. Removing the network file now has made it impossible for me to restart docker. So...yeah, it's pretty much a disaster. :-P

**Comment by kairok** [comment]
I did below command. It works for me.
systemctl stop docker;  systemctl start docker

**Comment by kt1024** [comment]
just happen, restart service/reboot host machine did not work. only docker service. no "docker-compose" used.

error logs:
```
Jul 06 12:50:22 x86 docker[13475]: docker: Error response from daemon: driver failed programming external connectivity on endpoint webdav (e083dcf
886d1f038115ec3d6f2cc0741d9650e0b011b998f596701398d427ceb): Bind for 0.0.0.0:7280 failed: port is already allocated.
```
```
# ss -tulnp | grep 7280
tcp     LISTEN   0        128                    *:7280                 *:*      users:(("docker-proxy",pid=819,fd=4)) 
```

```
# docker info
Client:
 Debug Mode: false

Server:
 Containers: 11
  Running: 10
  Paused: 0
  Stopped: 1
 Images: 38
 Server Version: 19.03.12
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.19.0-9-amd64
 Operating System: Debian GNU/Linux 10 (buster)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 31.33GiB
 Name: x86
 ID: OTHX:RFXS:3C3E:4CZW:Q4ZE:6ION:GB36:E75T:LVHE:QV52:WJ2U:PMCY
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support

```

**Comment by kt1024** [comment]
after delete local-kv.db and recreate all defined netorks, services is now back online.

**Comment by anvouk** [comment]
happened to me in a production on premise environment.... might as well add yet another possible way to solve this...

1. removed broken restarting container from compose
2. removed broken container from docker stopped list (docker ps -a)
3. reboot vm

hope this helps

**Comment by YassineHk** [comment]
it solved for me by deleting network 
`docker network disconnect -f {network-name} {endpoint-name}`
you can get {endpoint-name} form 
```
docker network ls
docker network {network-name} 
```

**Comment by Hipska** [comment]
Just had the same issue.
`docker info` said there were running containers while `docker ps -all` showed nothing. Restarting docker didn't solve this, proxy starts listening again on ports when restarted..

My solution; stop docker services, remove all under `/var/lib/docker/containers` start docker again and then docker compose..

**Comment by livengooddaily** [comment]
> Well, I've found a workaround:
> 
> * stop docker
> * remove all internal docker network: rm /var/lib/docker/network/files/
> * start docker

This solved the issue for me today, thanks.

**Comment by firecentaur** [comment]
> /var/lib/docker/network/files/

not a good idea. After I did this i got:
`Error response from daemon: failed to update store for object type *libnetwork.endpointCnt: Key not found in store`

**Comment by Cornul11** [comment]
> Well, I've found a workaround:
> 
> * stop docker
> * remove all internal docker network: rm /var/lib/docker/network/files/
> * start docker

more than five years later, this is still the only available workaround with no solution. That's interesting, to say the least.

**Comment by Roemer** [comment]
I have this issue with every reboot of my server. Why is this ticket closed when it seems still to be a problem?

**Comment by scottdotau** [comment]
> I have this issue with every reboot of my server. Why is this ticket closed when it seems still to be a problem?
> 
> Also using the workaround (deleting files from `/var/lib/docker/network/files/`) broke even more now. I now get `error="network some-id not found"`. I had to use `docker system prune` to delete all containers that could not start, recreate the needed networks and start all containers again.
> Next time I will try to just delete the PIDs of the docker-proxies which block the ports.

This issue still clearly exists - deleting above folder (mine was in a custom folder) did fix it, however I did have to recreate certain networks manually - which is to be expected considering my current setup (basically as if setting up for new).

Regardless, the fact that I needed to manually delete a system folder to get docker stable again (my rpi envs would just spam syslog errors til sys panic) isn't a great user experience and can potentially cause a lot of headaches.

The only reason I learned this was happening is because I have log2ram installed and an error in my logrotate config 🙃

**Comment by scottdotau** [comment]
My 250mb log2ram allocation would fill in ~15-20mins from a 75mb base after log rotations... From only 2-3 'ghost' containers. Not insignificant imo

**Comment by Hipska** [comment]
To me it didn't occur again after cleaning up `/var/lib/docker/containers`: https://github.com/moby/moby/issues/25981#issuecomment-1042901076

But I agree with others this is not a situation where the software should get in.

**Comment by scottdotau** [comment]
> @scottdotau If your still facing this issue, I'd need your `/var/lib/docker/network/files/local-kv.db` and all the files in `/var/lib/docker/containers/` to debug this further. Can you upload them here please? If you have some concerns about making these files publicly available, you can also DM them to me on [our community Slack](https://dockr.ly/slack) (I go by the same username).
> 
> Given your setup, I guess you don't have any logs from when you deleted these containers. Nonetheless, could you also check your logs to see if the daemon says anything when it starts?
> 
> EDIT: I forgot to also ask your `docker info`. Can you paste it here please?
> 
> @Hipska Are you still facing this issue?

Hey mate, have created archives of these in both my machines but the slack invite has expired ☺

Not sure the backups will help much because as you mentioned I rm -rf'd that shiz, but willing to help debug if if they do!

**Comment by akerouanton** [comment]
> Not sure the backups will help much because as you mentioned I rm -rf'd that shiz, but willing to help debug if if they do!

Yeah, unfortunately that won't help. I'd need those file from when the daemon is in a bad state.

**Comment by scottdotau** [comment]
> Yeah, unfortunately that won't help. I'd need those file from when the daemon is in a bad state.

Yeah np.

I have a good idea of how I can replicate it so will try do so over the weekend and save the logs.

**Comment by akerouanton** [comment]
That would be super helpful if you find a way to consistently reproduce this issue. If you can also put dockerd in debug mode ([see here](https://docs.docker.com/engine/reference/commandline/dockerd/#description)) before deleting containers, the daemon might also log interesting stuff.

**Comment by scottdotau** [comment]
> That would be super helpful if you find a way to consistently reproduce this issue. If you can also put dockerd in debug mode ([see here](https://docs.docker.com/engine/reference/commandline/dockerd/#description)) before deleting containers, the daemon might also log interesting stuff.

I *think*, based on my interpretation of the logs and what I was seeing when trying to figure it out, it isn't an issue with the containers themselves being able to start/stop but that the virtual networks those containers have/will create not being removed when those events happen, if that makes sense. 

This happened for me after a docker upgrade, lots of veth errors but I was too frustrated to pay too much attention 🙃

**Comment by scottdotau** [comment]
Hey @akerouanton,

I was able to reproduce this with a fairly small swarm stack & debugging on, however as mentioned unable to use the slack invite above and would prefer not uploading and linking.

If there's any other way I can get these through just let me know ☺.

**Comment by akerouanton** [comment]
@scottdotau Can you try this link please? https://join.slack.com/t/dockercommunity/shared_invite/zt-1v584f400-Bxd7sdEhCQSk75EE6Z03Sg

**Comment by fragmuffin** [comment]
Above workarounds didn't work for me, _including_ uninstalling docker and all dependencies, then re-installing.
What did work was manually removing all `/var/lib/docker` content using the script: [`nuke-graph-directory.sh`](https://github.com/moby/moby/blob/master/contrib/nuke-graph-directory.sh)
(not very useful for root cause, but may help other anxious engineers that land here)

**Comment by jansohn** [comment]
> [@scottdotau](https://github.com/scottdotau) Can you try this link please? https://join.slack.com/t/dockercommunity/shared_invite/zt-1v584f400-Bxd7sdEhCQSk75EE6Z03Sg

We've also run into this on one of our servers. Was there ever data uploaded for you to be analyzed? What's the status on this issue?

**Comment by mangococo** [comment]
I also encountered this problem on this version, but I really don't know the reproduction conditions.

When I tried to `kill` the useless `docker-proxy` processes, they turned into real zombie processes `[docker-proxy] <defunct>` . The problem disappeared after I restarted `dockerd`.

`docker version`

```
Client:
 Version:           20.10.24
 API version:       1.41
 Go version:        go1.19.7
 Git commit:        297e128
 Built:             Tue Apr  4 18:17:06 2023
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.24
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.19.7
  Git commit:       5d6db84
  Built:            Tue Apr  4 18:23:02 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.6.20
  GitCommit:        2806fc1057397dbaeefbea0e4e17bddfbd388f38
 runc:
  Version:          1.1.5
  GitCommit:        v1.1.5-0-gf19387a6
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Comment by robmry** [comment]
Hi @mangococo - 20.10.24 is not one of the [releases we can support](https://github.com/moby/moby/blob/master/project/BRANCHES-AND-TAGS.md) - upgrading to a current version is likely to fix it.

I don't see any reports of this issue against recent versions, and the code around port binding and starting/stopping docker-proxy has changed a lot over the last couple of years. I think this ticket probably covers several different issues with similar symptoms. 

So, I'll close it again.

If you still see the problem with a current version of Docker - please raise a new issue and include docker version + info output, and as much detail as possible about how we can repro.
